File: app/src/main/java/com/unidospelovolei/data/Mappers.kt
package com.unidospelovolei.data

import com.powersync.db.SqlCursor
import com.powersync.db.getBooleanOptional
import com.powersync.db.getDoubleOptional
import com.powersync.db.getLongOptional
import com.powersync.db.getString
import com.powersync.db.getStringOptional
import com.unidospelovolei.domain.model.Aviso
import com.unidospelovolei.domain.model.CategoriaPagina
import com.unidospelovolei.domain.model.Cobranca
import com.unidospelovolei.domain.model.ConfigFinanceiro
import com.unidospelovolei.domain.model.ConfigGrupo
import com.unidospelovolei.domain.model.Dica
import com.unidospelovolei.domain.model.Evento
import com.unidospelovolei.domain.model.Evolucao
import com.unidospelovolei.domain.model.Fundamento
import com.unidospelovolei.domain.model.Genero
import com.unidospelovolei.domain.model.MatchCard
import com.unidospelovolei.domain.model.MatchStatus
import com.unidospelovolei.domain.model.Pagamento
import com.unidospelovolei.domain.model.Pagina
import com.unidospelovolei.domain.model.Papel
import com.unidospelovolei.domain.model.Player
import com.unidospelovolei.domain.model.PlayerContato
import com.unidospelovolei.domain.model.Posicao
import com.unidospelovolei.domain.model.Post
import com.unidospelovolei.domain.model.Presenca
import com.unidospelovolei.domain.model.Regime
import com.unidospelovolei.domain.model.Standing
import com.unidospelovolei.domain.model.StatusPagamento
import com.unidospelovolei.domain.model.StatusPresenca
import com.unidospelovolei.domain.model.StatusVinculo
import com.unidospelovolei.domain.model.Team
import com.unidospelovolei.domain.model.TipoCobranca
import com.unidospelovolei.domain.model.TipoEvento
import com.unidospelovolei.domain.model.UserProfile
import com.unidospelovolei.domain.model.VinculoPedido

internal fun SqlCursor.int(
    name: String,
    padrao: Int = 0,
): Int = getLongOptional(name)?.toInt() ?: padrao

internal fun SqlCursor.bool(
    name: String,
    padrao: Boolean = false,
): Boolean = getBooleanOptional(name) ?: padrao

internal fun SqlCursor.intOrNull(name: String): Int? = getLongOptional(name)?.toInt()

internal fun SqlCursor.toPlayer(): Player =
    Player(
        id = getString("id"),
        nome = getStringOptional("nome").orEmpty(),
        skillLevel = int("skill_level", 3),
        genero = Genero.from(getStringOptional("genero")),
        ativo = bool("ativo", true),
        profileId = getStringOptional("profile_id"),
        posicao = Posicao.from(getStringOptional("posicao")),
        fotoUrl = getStringOptional("foto_url"),
        nascimentoDia = intOrNull("nascimento_dia"),
        nascimentoMes = intOrNull("nascimento_mes"),
        entrouEm = getStringOptional("entrou_em"),
        regime = Regime.from(getStringOptional("regime")),
    )

internal fun SqlCursor.realOrNull(name: String): Double? = getDoubleOptional(name)

internal fun SqlCursor.toPresenca(): Presenca =
    Presenca(
        id = getString("id"),
        playerId = getStringOptional("player_id").orEmpty(),
        data = getStringOptional("data").orEmpty(),
        status = StatusPresenca.from(getStringOptional("status")) ?: StatusPresenca.VOU,
        origem = getStringOptional("origem") ?: "atleta",
    )

internal fun SqlCursor.toAviso(): Aviso =
    Aviso(
        id = getString("id"),
        tipo = getStringOptional("tipo") ?: "lembrete",
        titulo = getStringOptional("titulo").orEmpty(),
        corpo = getStringOptional("corpo"),
        criadoEm = getStringOptional("criado_em"),
    )

internal fun SqlCursor.toConfigGrupo(): ConfigGrupo =
    ConfigGrupo(
        id = getString("id"),
        jogoHora = getStringOptional("jogo_hora") ?: "09:00",
        jogoLocal = getStringOptional("jogo_local"),
    )

internal fun SqlCursor.toPost(): Post =
    Post(
        id = getString("id"),
        autorNome = getStringOptional("autor_nome"),
        titulo = getStringOptional("titulo").orEmpty(),
        corpo = getStringOptional("corpo").orEmpty(),
        fixado = bool("fixado"),
        publicadoEm = getStringOptional("publicado_em"),
        reacoes = int("reacoes"),
        reagi = bool("reagi"),
    )

internal fun SqlCursor.toEvento(): Evento =
    Evento(
        id = getString("id"),
        titulo = getStringOptional("titulo").orEmpty(),
        descricao = getStringOptional("descricao"),
        tipo = TipoEvento.from(getStringOptional("tipo")),
        inicio = getStringOptional("inicio").orEmpty(),
        local = getStringOptional("local"),
    )

internal fun SqlCursor.toPagina(): Pagina =
    Pagina(
        id = getString("id"),
        slug = getStringOptional("slug").orEmpty(),
        categoria = CategoriaPagina.from(getStringOptional("categoria")),
        titulo = getStringOptional("titulo").orEmpty(),
        corpo = getStringOptional("corpo").orEmpty(),
        ordem = int("ordem"),
    )

internal fun SqlCursor.toConfigFinanceiro(): ConfigFinanceiro =
    ConfigFinanceiro(
        id = getString("id"),
        pixChave = getStringOptional("pix_chave"),
        pixNome = getStringOptional("pix_nome"),
        pixCidade = getStringOptional("pix_cidade"),
        mensalidadeCentavos = int("mensalidade_centavos"),
        diariaCentavos = int("diaria_centavos"),
    )

internal fun SqlCursor.toCobranca(): Cobranca =
    Cobranca(
        id = getString("id"),
        titulo = getStringOptional("titulo").orEmpty(),
        tipo = TipoCobranca.from(getStringOptional("tipo")),
        valorCentavos = int("valor_centavos"),
        competencia = getStringOptional("competencia"),
        venceEm = getStringOptional("vence_em"),
    )

internal fun SqlCursor.toPagamento(): Pagamento =
    Pagamento(
        id = getString("id"),
        cobrancaId = getStringOptional("cobranca_id").orEmpty(),
        playerId = getStringOptional("player_id").orEmpty(),
        valorCentavos = int("valor_centavos"),
        status = StatusPagamento.from(getStringOptional("status")),
        pagoEm = getStringOptional("pago_em"),
        observacao = getStringOptional("observacao"),
    )

internal fun SqlCursor.toEvolucao(): Evolucao {
    val medias =
        Fundamento.entries.mapNotNull { fundamento ->
            realOrNull("${fundamento.value}_media")?.let { fundamento to it }
        }
    return Evolucao(
        playerId = getStringOptional("player_id").orEmpty(),
        totalAvaliacoes = int("total_avaliacoes"),
        medias = medias.toMap(),
    )
}

internal fun SqlCursor.toDica(): Dica =
    Dica(
        id = getString("id"),
        fundamento = Fundamento.from(getStringOptional("atributo")) ?: Fundamento.ATITUDE,
        faixaMax = realOrNull("faixa_max") ?: 5.0,
        titulo = getStringOptional("titulo").orEmpty(),
        texto = getStringOptional("texto").orEmpty(),
    )

internal fun SqlCursor.toTeam(prefixo: String = ""): Team =
    Team(
        id = getString(prefixo + "id"),
        nome = getStringOptional(prefixo + "nome").orEmpty(),
        corHex = getStringOptional(prefixo + "cor_hex") ?: "#6B7280",
        sigla = getStringOptional(prefixo + "sigla").orEmpty(),
        ativo = bool(prefixo + "ativo", true),
        ordem = int(prefixo + "ordem"),
    )

internal fun SqlCursor.toMatchCard(): MatchCard =
    MatchCard(
        id = getString("id"),
        roundNumero = int("numero"),
        fase = int("fase", 1),
        quadra = int("quadra", 1),
        scoreA = int("score_a"),
        scoreB = int("score_b"),
        status = MatchStatus.from(getStringOptional("status")),
        winnerId = getStringOptional("winner_id"),
        teamA = toTeam("a_"),
        teamB = toTeam("b_"),
    )

internal fun SqlCursor.toStanding(): Standing =
    Standing(
        teamId = getString("team_id"),
        nome = getStringOptional("nome").orEmpty(),
        sigla = getStringOptional("sigla").orEmpty(),
        corHex = getStringOptional("cor_hex") ?: "#6B7280",
        jogos = int("jogos"),
        vitorias = int("vitorias"),
        derrotas = int("derrotas"),
        saldoPontos = int("saldo_pontos"),
        pontosPro = int("pontos_pro"),
    )

internal fun SqlCursor.toUserProfile(): UserProfile =
    UserProfile(
        id = getString("id"),
        email = getStringOptional("email"),
        nome = getStringOptional("nome"),
        papel =
            getStringOptional("papel")
                ?.let(Papel::from)
                ?: if (bool("is_admin")) Papel.DIRETORIA else Papel.ATLETA,
    )

internal fun SqlCursor.toVinculoPedido(): VinculoPedido =
    VinculoPedido(
        id = getString("id"),
        profileId = getStringOptional("profile_id").orEmpty(),
        playerId = getStringOptional("player_id").orEmpty(),
        profileNome = getStringOptional("profile_nome"),
        status = StatusVinculo.from(getStringOptional("status")),
        criadoEm = getStringOptional("criado_em"),
    )

internal fun SqlCursor.toPlayerContato(): PlayerContato =
    PlayerContato(
        id = getString("id"),
        playerId = getStringOptional("player_id").orEmpty(),
        telefone = getStringOptional("telefone"),
        contatoEmergencia = getStringOptional("contato_emergencia"),
        nascimentoAno = intOrNull("nascimento_ano"),
    )

File: web/src/data/mappers.ts
import {
  categoriaDe,
  EMOJI_PADRAO,
  fundamentoDe,
  FUNDAMENTOS,
  generoDe,
  papelDe,
  regimeDe,
  statusDe,
  statusPagamentoDe,
  statusPresencaDe,
  statusVinculoDe,
  tipoCobrancaDe,
  tipoEventoDe,
  type Cobranca,
  type ConfigFinanceiro,
  type ConfigGrupo,
  type Dica,
  type Evento,
  type Evolucao,
  type Fundamento,
  type MatchCard,
  type Pagamento,
  type Pagina,
  type Player,
  type PlayerContato,
  type Post,
  type Presenca,
  type Standing,
  type Team,
  type UserProfile,
  type VinculoPedido,
} from '../domain/models';

export type Row = Record<string, unknown>;

export const novoId = (): string =>
  typeof crypto !== 'undefined' && 'randomUUID' in crypto
    ? crypto.randomUUID()
    : `${Date.now()}-${Math.random().toString(16).slice(2)}`;

export const agoraIso = (): string => new Date().toISOString();

export const texto = (row: Row, nome: string): string => {
  const valor = row[nome];
  return typeof valor === 'string' ? valor : '';
};

export const textoOuNulo = (row: Row, nome: string): string | null => {
  const valor = row[nome];
  return typeof valor === 'string' && valor.length > 0 ? valor : null;
};

export const inteiro = (row: Row, nome: string, padrao = 0): number => {
  const valor = row[nome];
  return typeof valor === 'number' ? valor : padrao;
};

export const booleano = (row: Row, nome: string, padrao = false): boolean => {
  const valor = row[nome];
  if (typeof valor === 'number') return valor !== 0;
  if (typeof valor === 'boolean') return valor;
  return padrao;
};

export const inteiroOuNulo = (row: Row, nome: string): number | null => {
  const valor = row[nome];
  return typeof valor === 'number' ? valor : null;
};

export const toPlayer = (row: Row): Player => ({
  id: texto(row, 'id'),
  nome: texto(row, 'nome'),
  skillLevel: inteiro(row, 'skill_level', 3),
  genero: generoDe(textoOuNulo(row, 'genero')),
  ativo: booleano(row, 'ativo', true),
  profileId: textoOuNulo(row, 'profile_id'),
  fotoUrl: textoOuNulo(row, 'foto_url'),
  nascimentoDia: inteiroOuNulo(row, 'nascimento_dia'),
  nascimentoMes: inteiroOuNulo(row, 'nascimento_mes'),
  entrouEm: textoOuNulo(row, 'entrou_em'),
  regime: regimeDe(textoOuNulo(row, 'regime')),
});

export const toPresenca = (row: Row): Presenca => ({
  id: texto(row, 'id'),
  playerId: texto(row, 'player_id'),
  data: texto(row, 'data'),
  status: statusPresencaDe(textoOuNulo(row, 'status')) ?? 'vou',
  origem: textoOuNulo(row, 'origem') ?? 'atleta',
});

export const toConfigGrupo = (row: Row): ConfigGrupo => ({
  id: texto(row, 'id'),
  jogoHora: textoOuNulo(row, 'jogo_hora') ?? '09:00',
  jogoLocal: textoOuNulo(row, 'jogo_local'),
});

export const toPost = (row: Row): Post => ({
  id: texto(row, 'id'),
  autorNome: textoOuNulo(row, 'autor_nome'),
  titulo: texto(row, 'titulo'),
  corpo: texto(row, 'corpo'),
  fixado: booleano(row, 'fixado'),
  publicadoEm: textoOuNulo(row, 'publicado_em'),
  imagemUrl: textoOuNulo(row, 'imagem_url'),
  emoji: textoOuNulo(row, 'emoji')?.trim() || EMOJI_PADRAO,
  reacoes: inteiro(row, 'reacoes'),
  reagi: booleano(row, 'reagi'),
});

export const toEvento = (row: Row): Evento => ({
  id: texto(row, 'id'),
  titulo: texto(row, 'titulo'),
  descricao: textoOuNulo(row, 'descricao'),
  tipo: tipoEventoDe(textoOuNulo(row, 'tipo')),
  inicio: texto(row, 'inicio'),
  local: textoOuNulo(row, 'local'),
});

export const toPagina = (row: Row): Pagina => ({
  id: texto(row, 'id'),
  slug: texto(row, 'slug'),
  categoria: categoriaDe(textoOuNulo(row, 'categoria')),
  titulo: texto(row, 'titulo'),
  corpo: texto(row, 'corpo'),
  ordem: inteiro(row, 'ordem'),
});

export const toConfigFinanceiro = (row: Row): ConfigFinanceiro => ({
  id: texto(row, 'id'),
  pixChave: textoOuNulo(row, 'pix_chave'),
  pixNome: textoOuNulo(row, 'pix_nome'),
  pixCidade: textoOuNulo(row, 'pix_cidade'),
  mensalidadeCentavos: inteiro(row, 'mensalidade_centavos'),
  diariaCentavos: inteiro(row, 'diaria_centavos'),
});

export const toCobranca = (row: Row): Cobranca => ({
  id: texto(row, 'id'),
  titulo: texto(row, 'titulo'),
  tipo: tipoCobrancaDe(textoOuNulo(row, 'tipo')),
  valorCentavos: inteiro(row, 'valor_centavos'),
  competencia: textoOuNulo(row, 'competencia'),
  venceEm: textoOuNulo(row, 'vence_em'),
});

export const toPagamento = (row: Row): Pagamento => ({
  id: texto(row, 'id'),
  cobrancaId: texto(row, 'cobranca_id'),
  playerId: texto(row, 'player_id'),
  valorCentavos: inteiro(row, 'valor_centavos'),
  status: statusPagamentoDe(textoOuNulo(row, 'status')),
  pagoEm: textoOuNulo(row, 'pago_em'),
  observacao: textoOuNulo(row, 'observacao'),
});

export const toEvolucao = (row: Row): Evolucao => {
  const medias: Partial<Record<Fundamento, number>> = {};
  FUNDAMENTOS.forEach((fundamento) => {
    const valor = inteiroOuNulo(row, `${fundamento}_media`);
    if (valor !== null) medias[fundamento] = valor;
  });
  return {
    playerId: texto(row, 'player_id'),
    totalAvaliacoes: inteiro(row, 'total_avaliacoes'),
    medias,
  };
};

export const toDica = (row: Row): Dica => ({
  id: texto(row, 'id'),
  fundamento: fundamentoDe(textoOuNulo(row, 'atributo')) ?? 'atitude',
  faixaMax: inteiroOuNulo(row, 'faixa_max') ?? 5,
  titulo: texto(row, 'titulo'),
  texto: texto(row, 'texto'),
});

export const toTeam = (row: Row, prefixo = ''): Team => ({
  id: texto(row, `${prefixo}id`),
  nome: texto(row, `${prefixo}nome`),
  corHex: textoOuNulo(row, `${prefixo}cor_hex`) ?? '#6B7280',
  sigla: texto(row, `${prefixo}sigla`),
  ativo: booleano(row, `${prefixo}ativo`, true),
  ordem: inteiro(row, `${prefixo}ordem`),
});

export const toMatchCard = (row: Row): MatchCard => ({
  id: texto(row, 'id'),
  roundNumero: inteiro(row, 'numero'),
  fase: inteiro(row, 'fase', 1),
  quadra: inteiro(row, 'quadra', 1),
  scoreA: inteiro(row, 'score_a'),
  scoreB: inteiro(row, 'score_b'),
  status: statusDe(textoOuNulo(row, 'status')),
  winnerId: textoOuNulo(row, 'winner_id'),
  teamA: toTeam(row, 'a_'),
  teamB: toTeam(row, 'b_'),
});

export const toStanding = (row: Row): Standing => ({
  teamId: texto(row, 'team_id'),
  nome: texto(row, 'nome'),
  sigla: texto(row, 'sigla'),
  corHex: textoOuNulo(row, 'cor_hex') ?? '#6B7280',
  jogos: inteiro(row, 'jogos'),
  vitorias: inteiro(row, 'vitorias'),
  derrotas: inteiro(row, 'derrotas'),
  saldoPontos: inteiro(row, 'saldo_pontos'),
  pontosPro: inteiro(row, 'pontos_pro'),
});

export const toUserProfile = (row: Row): UserProfile => {
  const gravado = textoOuNulo(row, 'papel');
  const legado = booleano(row, 'is_admin') ? 'diretoria' : 'atleta';
  const resolvido = papelDe(gravado ?? legado);
  return {
    id: texto(row, 'id'),
    email: textoOuNulo(row, 'email'),
    nome: textoOuNulo(row, 'nome'),
    papel: resolvido,
    isAdmin: resolvido === 'diretoria',
  };
};

export const toVinculoPedido = (row: Row): VinculoPedido => ({
  id: texto(row, 'id'),
  profileId: texto(row, 'profile_id'),
  playerId: texto(row, 'player_id'),
  profileNome: textoOuNulo(row, 'profile_nome'),
  status: statusVinculoDe(textoOuNulo(row, 'status')),
  criadoEm: textoOuNulo(row, 'criado_em'),
});

export const toPlayerContato = (row: Row): PlayerContato => ({
  id: texto(row, 'id'),
  playerId: texto(row, 'player_id'),
  telefone: textoOuNulo(row, 'telefone'),
  contatoEmergencia: textoOuNulo(row, 'contato_emergencia'),
  nascimentoAno: inteiroOuNulo(row, 'nascimento_ano'),
});
